File: 022_generate_parentheses.java
class Solution {
    public List<String> generateParenthesis(int n) {
        List<String> result = new ArrayList<>();
        backtrace(result, "", 0, 0, n);
        return result;
    }

    private void backtrace(List<String> result, String curr, int open, int close, int max) {
        if (curr.length() == 2 * max) {
            result.add(curr);
            return;
        }

        if (open < max)
            backtrace(result, curr + "(", open + 1, close, max);
        if (close < open)
            backtrace(result, curr + ")", open, close + 1, max);
    }
}

File: python/022_generate_parentheses.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        res = []
        self.backtrace(n, 0, 0, "", res)
        return res

    def backtrace(self, n, open_count, close_count, ans: str, res: List[str]):
        if len(ans) == 2 * n:
            res.append(ans)
            return

        if open_count < n:
            self.backtrace(n, open_count + 1, close_count, ans + "(", res)
        if close_count < open_count:
            self.backtrace(n, open_count, close_count + 1, ans + ")", res)
